Add PG reserved buffer test for QoS buffer management

prasoon-nexthop opened the pull request:
This commit adds a new test to validate PG (Priority Group) reserved buffer behavior for QoS buffer management on TH5 for x86_64_nexthop* DUTs.

Test Overview:
- Creates 2 buffer profiles for 2 different PGs (PG0 and PG1)
- PG0: No PG MIN (uses shared pool only)
- PG1: Has PG MIN threshold configured
- Validates that PG0 traffic exhausts shared pool and drops excess packets
- Validates that PG1 gets guaranteed bandwidth via PG MIN

Changes:
1. tests/saitests/py3/sai_qos_tests.py: Added PgMinThresholdTest class
2. tests/qos/test_qos_sai.py: Added testQosSaiPgMinThreshold() method
3. tests/qos/qos_sai_base.py: Added skip_pcap parameter to avoid OOM
4. tests/qos/files/qos_params.th5.yaml: Added pg_min_threshold config
5. tests/common/plugins/conditional_mark/tests_mark_conditions.yaml: Added conditional marker here

Tests can be enhanced to work for other vendors using Broadcom's TH5 based DUTs.

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [x] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] New Test case
    - [ ] Skipped for non-supported platforms
- [x] Test case improvement


### Back port request
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405
- [ ] 202411
- [x] 202505

### Approach
#### What is the motivation for this PR?

#### How did you do it?

#### How did you verify/test it?

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

prasoon-nexthop commented:
Tagging @StormLiangMS

prasoon-nexthop commented:
@vmittal-msft 
Vineet, can this be reviewed please?

prasoon-nexthop commented on tests/saitests/py3/sai_qos_tests.py:
```diff
@@ -7344,3 +7344,236 @@ def runTest(self):
+                print("Assert 3 WARNING: PG1 (lossless) has more drops than expected ({} > {})".format(
+                    pg1_drops, self.margin), file=sys.stderr)
+
+            # Assert 4: Lossy should drop more than or equal to lossless (main validation)
```
Right, the assertion validates that PFC is working correctly - if lossless traffic drops as much as lossy, it means PFC is broken or misconfigured.

vmittal-msft replied:
@prasoon-nexthop  Please let me know if my understanding is correct. In this case, we expect lossy to drop once it crosses shared pool limit and lossless to drop/slow down once it hits min_pg limit. So if the pool is filled with lossy till shared limit, then lossless may just fill till min_pg limit and then it should start sending pfc to source to slow down since no space for lossless in shared space. so in this case, lossless may drop few or zero and lossy after shared limit reach. is this correct ?

prasoon-nexthop replied:
Hi @vmittal-msft 
Your understanding is mostly correct: Lossy drops when shared pool is exhausted, lossless uses reserved buffer and triggers PFC to avoid drops. The refinement is that PFC is triggered by XOFF threshold (not just min_pg limit), and it's the upstream sender that slows down (not the lossless queue itself). The test validates that this mechanism works: lossy drops ≥ lossless drops.

prasoon-nexthop commented:
@vmittal-msft Can this be merged too? I see two pending reviews.
Appreciate your help on this.

prasoon-nexthop commented:
> > @vmittal-msft Can this be merged too? I see two pending reviews. Appreciate your help on this.
> 
> @prasoon-nexthop I would wait for other reviewers to approve as well @kperumalbfn @bingwang-ms. Also since this is applicable for TH5 only, can we put check so this runs for TH5 only and not cause failure for others ?

Actually the "pg_min_threshold:" section in the TH5 yaml file will ensure that this only runs on TH5 else the test will be skipped


```
        if "pg_min_threshold" not in qosConfig:
            pytest.skip("PG MIN threshold test parameters not configured for this platform")
            
```

prasoon-nexthop commented:
@bingwang-ms @kperumalbfn 

Requesting review on this.

prasoon-nexthop commented:
Awaiting response on this

@vmittal-msft @bingwang-ms @kperumalbfn

kperumalbfn commented on tests/qos/files/qos_params.th5.yaml:
```diff
@@ -3,6 +3,20 @@ qos_params:
+                shared_pool_size: 166000000
+                pg1_min_size: 10000000
+                pg0_pkts_to_fill: 2000000
+                pg1_pkts_to_fill: 2000000
```
@prasoon-nexthop Thanks for adding this. Could you also add test for 2 lossless PGs. This would need a config change, but will be good to add.

kperumalbfn replied:
Thanks, should be good. Please add a follow-up sonic-mgmt PR to cover 2 lossless PGs

prasoon-nexthop replied:
Sure, thanks @kperumalbfn 
Once this PR gets merged, I will raise another handling other scenarios.

prasoon-nexthop commented on tests/saitests/py3/sai_qos_tests.py:
```diff
@@ -7344,3 +7344,236 @@ def runTest(self):
+            self.wait_for_counter_update(timeout=10, interval=0.5)
+
+            # Read drop counters and packet counters
+            pg_drop_counters = sai_thrift_read_pg_drop_counters(
```
Hi @kperumalbfn 

This test is specifically designed for and limited to TH (Tomahawk) platforms, which support per-PG drop counters via flex counters (BRCM_SAI_IS_FEAT_INGRESS_FLEXCTR_TC_DROP).

The test is gated by SUPPORTED_ASIC_LIST at the top of the file, which only includes ASICs 
that have been verified to support SAI_INGRESS_PRIORITY_GROUP_STAT_DROPPED_PACKETS.

For ASICs that don't support per-PG drop counters they won't be in SUPPORTED_ASIC_LIST, so the test will be skipped.

If we want to extend this test to ASICs without per-PG drop counter support, we would need to:
- Add a feature detection check (e.g., check if SAI returns SAI_STATUS_NOT_SUPPORTED)
- Fall back to port-level ingress drop counters (SAI_PORT_STAT_IF_IN_DISCARDS)
- Adjust validation logic since port-level counters aggregate all PGs

prasoon-nexthop commented:
> can we please ensure the test case itself is asic agnostic?

Hi Rita

The test itself is pretty much agnostic of the ASIC,  we have tested on TH based platforms and are in the process of enabling it for DNX etc.  In case it needs to be enabled for other vendors there may be minor changes needed in the yaml file to add pg_min_threshold etc.

prasoon-nexthop commented:
> do we have same coverage for multi-asic device too?

Hi Rita

The test already uses the `get_src_dst_asic_and_duts` fixture, which is the standard mechanism for multi-ASIC support in QoS SAI tests. Looking at the code:


```
def testQosSaiPgMinThreshold(
    self, ptfhost, dutTestParams, dutConfig, dutQosConfig,
    get_src_dst_asic_and_duts  # ← Already has multi-ASIC fixture!
):
    # ...
    src_dut_index = get_src_dst_asic_and_duts['src_dut_index']
    src_asic_index = get_src_dst_asic_and_duts['src_asic_index']
    dst_dut_index = get_src_dst_asic_and_duts['dst_dut_index']
    dst_asic_index = get_src_dst_asic_and_duts['dst_asic_index']
    
    src_testPortIps = dutConfig["testPortIps"][src_dut_index][src_asic_index]
    dst_testPortIps = dutConfig["testPortIps"][dst_dut_index][dst_asic_index]
    
```

For multi-asic duts the indices are automatically set as the following

```
            dst_dut_index = src_dut_index
            src_asic_index = 0
            dst_asic_index = 1
            
```

We don't have mulit-asic duts available in house but if they are available the test will automatically

- Detect multi-ASIC DUTs via duthost.sonichost.is_multi_asic
- Select appropriate ASICs based on topology

Only thing needed would be the following

```
@pytest.mark.parametrize("test_port_selection_criteria", 
                         ["single_asic", "single_dut_multi_asic"])
def testQosSaiPgMinThreshold(
...
)
```

prasoon-nexthop commented:
> @prasoon-nexthop @kperumalbfn @ansrajpu-git thanks for joining meeting. Few action items -
> 
> 1. Please have links of follow up Issues/PRs e.g. Q3D etc.
> 2. Put a conditional marker to skip this test since flex counter support for PG drops is not enabled by default
> 3. Please update "PG min" to "PG reserved buffer"
> 4. @ansrajpu-git  to check if DNX support per PG drop counter today
> 5. We can have PG drop as well as PG min test separately in qos test suite
> 
> Please add if i missed anything.

Done. Added a conditional marker to skip this test on other platforms. 
Also modified the description.

prasoon-nexthop commented:
Here is the ticket tracking this https://github.com/sonic-net/sonic-mgmt/issues/21611

mssonicbld commented:
/azp run

kperumalbfn commented:
@prasoon-nexthop Could you update the PR description with the latest change.

prasoon-nexthop commented:
> @prasoon-nexthop are we handling Q3D as separate PR ?

Thats correct, Vineet.
